### DISCUSSION TITLE
This is not possible to delete stacks with name longer than 128 chars

### DISCUSSION THREAD

**Discussion by ikhvostenkov** [aws/aws-cli#6847] [Q&A] [answered]
### Describe the bug

This is possible to create stack with name longer than 128 chars through _CDK_, but this is not possible to delete stack with _AWS cli_ with stack name longer than 128 chars. Probably this should be prohibited to create stacks with name longer than 128 char in _CDK_, but as soon as this is possible you should have chance to delete it, especially if you have ~100 such stacks, doing this through AWS Console manually is annoying. 

### Expected Behavior

Stack is successfully deleted.

### Current Behavior

An error occurred (ValidationError) when calling the DeleteStack operation: Stack name cannot exceed 128 characters

### Steps to Reproduce:
1. Create any CFN stack with name longer than 128 characters, not from AWS console, but using _cdk_ construct. This is possible to do it with _cdk_ constructs.
2. Try to delete stack with _aws cli_:
```aws cloudformation delete-stack --stack-name {name} ```

### CLI version used

 aws --version aws-cli/1.20.6 Python/3.8.6 Darwin/21.3.0 botocore/1.21.7

### Environment details (OS name and version, etc.)

sw_vers ProductName:	macOS ProductVersion:	12.2.1 BuildVersion:	21D62

**Comment by kdaily** [comment, marked as answer]
Hi @ikhvostenkov,

Thanks for writing. According to the [CloudFormation `CreateStack` documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CreateStack.html), stack names are limited to 128 characters:

> StackName
> 
>     The name that's associated with the stack. The name must be unique in the Region in which you are creating the stack.
> Note
> 
> A stack name can contain only alphanumeric characters (case sensitive) and hyphens. It must start with an alphabetical character and can't be longer than 128 characters.

It looks like that would be a server side Validation error (if you provide `--debug` logs we can confirm), as the API does not model the length limitations for us to do validation on client side.

I'll check in with the CDK team for some more information.

**Comment by kellertk** [comment]
While it appears that CDK does [not check for this at synthesis time,](https://github.com/aws/aws-cdk/blob/9457471f92ec11074166bb12fc12edd86c475261/packages/%40aws-cdk/core/lib/stack.ts#L373), I wasn't able to get a test stack deployed with a really long name:

```typescript
#!/usr/bin/env node
import * as cdk from 'aws-cdk-lib';
import { CdkTestStack } from '../lib/cdk-test-stack';

const app = new cdk.App();
new CdkTestStack(app, 'LookAtMyReallyLongStackNameThisStackNameIsLongerThan128CharactersThatIsNutsIDontThinkThereIsEnoughAWSAvailableToLetEveryoneHaveStackNamesThisLong');
```

```powershell
PS C:\Users\theto\Documents\GitHub\cdk-test> cdk deploy   

✨  Synthesis time: 8.85s


Stack name cannot exceed 128 characters
PS C:\Users\theto\Documents\GitHub\cdk-test>
```
It will actually `synth` fine though. 🤔 

Edit: After some digging it looks like this error is coming right from the CloudFormation service API.

**Comment by peterwoodworth** [comment]
@ikhvostenkov I'm curious how you initially deployed the stacks (i.e. `cdk deploy`), and how long ago?

**Comment by ikhvostenkov** [comment]
@kdaily @peterwoodworth @kellertk  Sorry, I confused you. I just checked the code and these stacks were created with JAVA SDK, not with cdk, by some reason I thought these stacks were created with cdk. Here is the code snippet:
```
createStackRequest = CreateStackRequest.builder()
            .stackName({NAME})
            .templateBody({BODY})
            .roleARN({ROLE})
            .parameters(parameters)
            .build()

        cloudFormationClient.createStack(createStackRequest)
        ```

These stacks were created ~ 6 months ago. I dropped everything, I can reproduce it later if required.

**Comment by kdaily** [comment]
Thanks @ikhvostenkov! I'll check with our Java team too, but since the error comes from the service, the Java SDK should now also error with the same response.

**Comment by ikhvostenkov** [comment]
@kdaily I totally agree, that this is a bug in Java SDK, however, I think, this should be possible to delete any stack despite the stack's name length via AWS cli, as that might be other people having the same problem. And also this is possible to delete stack with name longer than 128 chars via AWS Console and then there is discrepancy between how AWS Console works and how AWS cli works.

**Comment by kdaily** [comment]
Hi @ikhvostenkov,

I'm checked with the Java team, and supplying a stack name greater than 128 characters fails with the following error, which is the same server side error received through the AWS CLI:

```
software.amazon.awssdk.services.cloudformation.model.CloudFormationException: Stack name cannot exceed 128 characters (Service: CloudFormation, Status Code: 400, Request ID: e4ae2cb4-2433-4654-9baa-ef7b0a618bdd)
```

I also tried to create a stack with the AWS console, which failed as well:

<img width="1630" alt="Screen Shot 2022-04-04 at 1 06 13 PM" src="https://user-images.githubusercontent.com/84993/161625718-b89a388f-d530-4310-a92a-327254ff7d50.png">

So, I think I'll need to see how you can reproduce this issue with a longer stack name!

**Comment by kdaily** [comment]
Just a heads up that I'm going to migrate this question to the AWS CLI GitHub discussions to continue our conversation. Thanks!
